src/dodal/devices/oav/pin_tip_detection/oav_with_pin_tip_detection.py

```diff
@@ -0,0 +1,208 @@
+
+    async def connect(self, sim: bool = False):
+        for signal in [self.array_data, self.oav_width, self.oav_height]:
+            await signal.connect(sim=sim)
+
+        for soft_signal in [
+            self.timeout,
+            self.preprocess,
+            self.canny_upper,
+            self.canny_lower,
+            self.close_ksize,
+            self.close_iterations,
+            self.scan_direction,
+            self.min_tip_height,
+        ]:
+            await soft_signal.connect(sim=False)  # Soft signals don't need simulating.
+
+    async def _get_tip_position(
+        self,
+    ) -> Tuple[Tuple[Optional[int], Optional[int]], float]:
+        """
+        Gets the location of the pin tip.
+
+        Returns tuple of:
+            ((tip_x, tip_y), timestamp)
+        """
+        sample_detection = MxSampleDetect(
+            preprocess=await self.preprocess.get_value(),
+            canny_lower=await self.canny_lower.get_value(),
+            canny_upper=await self.canny_upper.get_value(),
+            close_ksize=await self.close_ksize.get_value(),
+            close_iterations=await self.close_iterations.get_value(),
+            scan_direction=await self.scan_direction.get_value(),
+            min_tip_height=await self.min_tip_height.get_value(),
+        )
+
+        array_reading: dict[str, Reading] = await self.array_data.read()
+        array_data: NDArray[np.uint8] = array_reading[""]["value"]
+        timestamp: float = array_reading[""]["timestamp"]
+
+        height: int = await self.oav_height.get_value()
+        width: int = await self.oav_width.get_value()
+
+        num_pixels: int = height * width  # type: ignore
+        value_len = array_data.shape[0]
+
+        try:
+            if value_len == num_pixels * 3:
+                # RGB data
+                value = array_data.reshape(height, width, 3)
+            elif value_len == num_pixels:
+                # Grayscale data
+                value = array_data.reshape(height, width)
+            else:
+                # Something else?
+                raise ValueError(
+                    "Unexpected data array size: expected {} (grayscale data) or {} (rgb data), got {}",
+                    num_pixels,
+                    num_pixels * 3,
+                    value_len,
+                )
+
+            start_time = time.time()
+            location = sample_detection.processArray(value)
+            end_time = time.time()
+            LOGGER.debug(
+                "Sample location detection took {}ms".format(
+                    (end_time - start_time) * 1000.0
+                )
+            )
+            tip_x = location.tip_x
+            tip_y = location.tip_y
+        except Exception as e:
+            LOGGER.error("Failed to detect pin-tip location due to exception: ", e)
+            tip_x = None
+            tip_y = None
+
+        return (tip_x, tip_y), timestamp
+
+    async def read(self) -> dict[str, Reading]:
```

Comment:
Should: There are times that we may want the whole edge array (for grid detection). We can do this in a new issue if you like but we will need it

Reply:
New issue at https://github.com/DiamondLightSource/dodal/issues/146